package/samplers/pfns4bo/sampler.py

```diff
@@ -0,0 +1,274 @@
+        "efficient_eval_masking": True,
+        "nlayers": 12,
+    }
+
+    bs = 128
+    all_targets = []
+    for num_hps in [
+        2,
+        8,
+        12,
+    ]:  # a few different samples in case the number of features makes a difference in y dist
+        b = config["priordataloader_class_or_get_batch"].get_batch_method(
+            bs,
+            1000,
+            num_hps,
+            epoch=0,
+            device=device,
+            hyperparameters={
+                **config["extra_prior_kwargs_dict"]["hyperparameters"],
+                "num_hyperparameter_samples_per_batch": -1,
+            },
+        )
+        all_targets.append(b.target_y.flatten())
+    ys = torch.cat(all_targets, 0).cpu()
+
+    config["criterion"] = bar_distribution.FullSupportBarDistribution(
+        bar_distribution.get_bucket_limits(1000, ys=ys)
+    )
+
+    return config
+
+
+class PFNs4BOSampler(BaseSampler):
+    def __init__(
+        self,
+        *,
+        prior: str | torch.nn.Module = "hebo",
+        model_path: str | None = None,
+        seed: int | None = None,
+        independent_sampler: BaseSampler | None = None,
+        n_startup_trials: int = 10,
+        num_grad_steps: int = 15_000,
+        num_candidates: int = 100,
+        pre_sample_size: int = 100_000,
+    ) -> None:
+        self._num_grad_steps = num_grad_steps
+        self._num_candidates = num_candidates
+        self._pre_sample_size = pre_sample_size
+
+        self._rng = LazyRandomState(seed)
+        self._independent_sampler = independent_sampler or RandomSampler(seed=seed)
+        self._intersection_search_space = IntersectionSearchSpace()
+        self._n_startup_trials = n_startup_trials
+
+        self._device = utils.default_device
+
+        if isinstance(prior, torch.nn.Module):
+            trained_model = prior
+        elif prior == "vanilla gp":
+            _, _, trained_model, _ = train(**get_vanilla_gp_config(self._device))
+        elif prior == "hebo":
+            _, _, trained_model, _ = train(**get_heboplus_config(self._device))
+        else:
+            raise ValueError("You should specify `prior` as 'vanilla gp', 'hebo', or a model.")
+
+        self._model = trained_model
+        self._model.eval()
+
+        if model_path is not None:
+            torch.save(trained_model, model_path)
+
+    def sample_relative(
+        self, study: Study, trial: Trial, search_space: dict[str, BaseDistribution]
+    ) -> dict[str, Any]:
+        self._raise_error_if_multi_objective(study)
+
+        if search_space == {}:
+            return {}
+
+        states = (TrialState.COMPLETE,)
+        trials = study._get_trials(deepcopy=False, states=states, use_cache=True)
+
+        if len(trials) < self._n_startup_trials:
+            return {}
+
+        (
+            internal_search_space,
+            normalized_params,
+        ) = gp_search_space.get_search_space_and_normalized_params(trials, search_space)
```

Comment:
The normalization of the HPs should be to [0,1] interval. Not sure, if that is what this function does. Just as a note :)

Reply:
Thanks for the comment. Yes, this search space transformation allow us to transform each variable into [0,1] interval.